### PR TITLE
feat: Sliding Canvas for Dragging and Selection

### LINKED ISSUE #8327: [Bug] Jitters and problems with drag and drop
## Description
Creating this issue to list current problems with the new drag and drop library. I think the issues are scattered so creating a new one to collate them.

1. Issues in dragging/resizing upwards
2. Jitters in container and iframe specifically(reported on Discord)
3. Drag and drop sometimes unresponsive when you try to draw a polygon from outside the canvas area(difficult to consistently reproduce this)

### LINKED ISSUE #10195: [Bug]: Widget inside container widgets are added in a different alignment when there is a space in canvas(towards bottom)
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

[LOOM DEMO](https://www.loom.com/share/ee10b826feca4a49bd7c3a10740a149c)

### Steps To Reproduce

1. Add container widget
2. scroll the canvas, so that container widget is not completely visible in canvas
3. Add any widget inside the container widget
4. Observe there is alignment issue while adding widget inside container

### Environment

Release

### Version

Cloud

### PR DESCRIPTION
**Pull request by marks0351**
> Pull Request Template
>
> Use this template to quickly create a well written pull request. Delete all quotes before creating the pull request.

## Description

Appsmith DnD improvements.
using ResizeObserver and IntersectionObserver to implement sliding canvas properly.
fixing few DnD fixes as well.

Fixes #8327
Fixes #10195

> if no issue exists, please create an issue and ask the maintainers about this first

## Type of change

> Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

## How Has This Been Tested?

> Please describe the tests that you ran to verify your changes. Provide instructions, so we can reproduce.
> Please also list any relevant details for your test configuration.

- Test A
- Test B

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: feat/sliding-canvas 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 55.11 **(0)** | 36.92 **(0.05)** | 34.99 **(0)** | 55.59 **(0)**
 :green_circle: | app/client/src/components/editorComponents/DropTargetUtils.ts | 100 **(10)** | 66.67 **(0)** | 100 **(0)** | 100 **(11.11)**
 :red_circle: | app/client/src/pages/Editor/MultiSelectPropertyPane/index.tsx | 85 **(-5)** | 50 **(-50)** | 66.67 **(0)** | 88.24 **(0)**
 :sparkles: :new: | **app/client/src/pages/common/CanvasArenas/CanvasDraggingArena.tsx** | **100** | **87.5** | **100** | **100**
 :sparkles: :new: | **app/client/src/pages/common/CanvasArenas/CanvasMultiPointerArena.tsx** | **25.61** | **5.88** | **0** | **26.58**
 :sparkles: :new: | **app/client/src/pages/common/CanvasArenas/CanvasSelectionArena.tsx** | **93.72** | **77.08** | **88.89** | **93.58**
 :sparkles: :new: | **app/client/src/pages/common/CanvasArenas/StickyCanvasArena.tsx** | **65.38** | **71.43** | **54.55** | **63.27**
 :sparkles: :new: | **app/client/src/pages/common/CanvasArenas/hooks/useBlocksToBeDraggedOnCanvas.ts** | **92.59** | **87.21** | **83.33** | **93.75**
 :sparkles: :new: | **app/client/src/pages/common/CanvasArenas/hooks/useCanvasDragToScroll.ts** | **97.67** | **88.37** | **100** | **97.62**
 :sparkles: :new: | **app/client/src/pages/common/CanvasArenas/hooks/useCanvasDragging.ts** | **85.15** | **67.3** | **83.87** | **85.03**
 :sparkles: :new: | **app/client/src/sagas/CanvasSagas/DraggingCanvasSagas.ts** | **76.47** | **53.85** | **75** | **78.02**
 :sparkles: :new: | **app/client/src/sagas/CanvasSagas/SelectionCanvasSagas.ts** | **97.83** | **86.36** | **87.5** | **97.56**
 :red_circle: | app/client/src/utils/WorkerUtil.ts | 89.76 **(0)** | 70.59 **(-1.96)** | 100 **(0)** | 93.33 **(0)**
 :green_circle: | app/client/src/utils/helpers.tsx | 63.46 **(0.34)** | 30 **(0.56)** | 50.98 **(0)** | 60.68 **(0)**
 :green_circle: | app/client/src/widgets/ChartWidget/component/index.tsx | 50.59 **(1.18)** | 23.71 **(1.03)** | 53.13 **(3.13)** | 48.1 **(0.63)**
 :x: | ~~app/client/src/pages/common/CanvasDraggingArena.tsx~~ | ~~100~~ | ~~90~~ | ~~100~~ | ~~100~~
 :x: | ~~app/client/src/pages/common/CanvasMultiPointerArena.tsx~~ | ~~25.61~~ | ~~5.88~~ | ~~0~~ | ~~26.58~~
 :x: | ~~app/client/src/pages/common/CanvasSelectionArena.tsx~~ | ~~91.58~~ | ~~72.13~~ | ~~89.29~~ | ~~91.3~~
 :x: | ~~app/client/src/sagas/DraggingCanvasSagas.ts~~ | ~~76~~ | ~~53.85~~ | ~~72.73~~ | ~~77.78~~
 :x: | ~~app/client/src/sagas/SelectionCanvasSagas.ts~~ | ~~100~~ | ~~72.22~~ | ~~100~~ | ~~100~~
 :x: | ~~app/client/src/utils/hooks/useBlocksToBeDraggedOnCanvas.ts~~ | ~~92.48~~ | ~~87.21~~ | ~~82.76~~ | ~~93.7~~
 :x: | ~~app/client/src/utils/hooks/useCanvasDragToScroll.ts~~ | ~~97.67~~ | ~~90.7~~ | ~~100~~ | ~~97.62~~
 :x: | ~~app/client/src/utils/hooks/useCanvasDragging.ts~~ | ~~86.79~~ | ~~68.29~~ | ~~87.88~~ | ~~86.73~~</details>

### CODE REVIEW

File: app/client/src/pages/common/CanvasArenas/hooks/useCanvasDragging.ts [hunk 1]
```diff
@@ -645,14 +590,20 @@ export const useCanvasDragging = (
...
+            stickyCanvasRef.current &&
+            slidingArenaRef.current &&
             scrollParent &&
             isCurrentDraggedCanvas &&
             canvasIsDragging
           ) {
-            const canvasCtx: any = canvasDrawRef.current.getContext("2d");
-            const topOffset = getCanvasTopOffset();
+            const canvasCtx: any = stickyCanvasRef.current.getContext("2d");
+            stickyCanvasRef.current.style.top;
```
Comment (rahulramesha) [inline app/client/src/pages/common/CanvasArenas/hooks/useCanvasDragging.ts:600]: not sure what this line is doing here

**Review by rahulramesha — APPROVED** [review]

**Comment by vivekverma2312** [comment]
@marks0351 Tested the deploy preview and working fine.

**Pull request merged**